From acc363eac9885bcfce891d64697d863b88277619 Mon Sep 17 00:00:00 2001
From: Olivier Brunel <jjk@jjacky.com>
Date: Thu, 16 Jan 2014 17:44:31 +0100
Subject: [PATCH 2/2] menu: Add helper gtk_menu_popup_and_destroy()

Often one will create a menu in order to pop it up once and be done with
it. In such cases, they'd need not to forget to destroy it on popdown
to avoid a leak.

This helper makes it convenient to do so, which can also serves as a
reminder that menus used with gtk_menu_popup() needs to either be
manually destroyed, or ref_sink/unref-ed to avoid leaks.

https://bugzilla.gnome.org/show_bug.cgi?id=720401
Signed-off-by: Olivier Brunel <jjk@jjacky.com>
---
 docs/reference/gtk/gtk3-sections.txt |  1 +
 gtk/gtkmenu.c                        | 39 ++++++++++++++++++++++++++++++++++++
 gtk/gtkmenu.h                        |  6 ++++++
 3 files changed, 46 insertions(+)

diff --git a/docs/reference/gtk/gtk3-sections.txt b/docs/reference/gtk/gtk3-sections.txt
index e2d0435..75e2923 100644
--- a/docs/reference/gtk/gtk3-sections.txt
+++ b/docs/reference/gtk/gtk3-sections.txt
@@ -2217,6 +2217,7 @@ gtk_menu_reorder_child
 gtk_menu_attach
 gtk_menu_popup_for_device
 gtk_menu_popup
+gtk_menu_popup_and_destroy
 gtk_menu_set_accel_group
 gtk_menu_get_accel_group
 gtk_menu_set_accel_path
diff --git a/gtk/gtkmenu.c b/gtk/gtkmenu.c
index f622855..c202b0e 100644
--- a/gtk/gtkmenu.c
+++ b/gtk/gtkmenu.c
@@ -1767,6 +1767,10 @@ gtk_menu_popup_for_device (GtkMenu             *menu,
  * a mouse click or key press) that caused the initiation of the popup.
  * Only if no such event is available, gtk_get_current_event_time() can
  * be used instead.
+ *
+ * Note that the menu isn't destroyed on popdown, you need to call
+ * gtk_widget_destroy() yourself. If you only intend to popup the menu once and
+ * never use it again, you can use gtk_menu_popup_and_destroy()
  */
 void
 gtk_menu_popup (GtkMenu             *menu,
@@ -1788,6 +1792,41 @@ gtk_menu_popup (GtkMenu             *menu,
 }
 
 /**
+ * gtk_menu_popup_and_destroy:
+ * @menu: a #GtkMenu
+ * @func: (scope async) (allow-none): a user supplied function used to position
+ *     the menu, or %NULL
+ * @data: user supplied data to be passed to @func.
+ * @button: the mouse button which was pressed to initiate the event.
+ * @activate_time: the time at which the activation event occurred.
+ *
+ * Displays a menu and makes it available for selection.
+ *
+ * This is similar to gtk_menu_popup() except that @menu will be destroyed when
+ * hidden. Useful e.g. for context-sensitive menus you want to only popup once
+ * and don't need to keep around afterwards.
+ *
+ * Since: 3.12
+ */
+void
+gtk_menu_popup_and_destroy (GtkMenu             *menu,
+                            GtkMenuPositionFunc  func,
+                            gpointer             data,
+                            guint                button,
+                            guint32              activate_time)
+{
+  g_return_if_fail (GTK_IS_MENU (menu));
+
+  g_signal_connect_swapped (menu, "hide", G_CALLBACK (gtk_widget_destroy), menu);
+  gtk_menu_popup_for_device (menu,
+                             NULL,
+                             NULL,
+                             NULL,
+                             func, data, NULL,
+                             button, activate_time);
+}
+
+/**
  * gtk_menu_popdown:
  * @menu: a #GtkMenu
  *
diff --git a/gtk/gtkmenu.h b/gtk/gtkmenu.h
index bf81a89..e996bf1 100644
--- a/gtk/gtkmenu.h
+++ b/gtk/gtkmenu.h
@@ -144,6 +144,12 @@ void	   gtk_menu_popup		  (GtkMenu	       *menu,
 					   gpointer		data,
 					   guint		button,
 					   guint32		activate_time);
+GDK_AVAILABLE_IN_3_12
+void        gtk_menu_popup_and_destroy    (GtkMenu             *menu,
+                                           GtkMenuPositionFunc  func,
+                                           gpointer             data,
+                                           guint                button,
+                                           guint32              activate_time);
 GDK_AVAILABLE_IN_ALL
 void       gtk_menu_popup_for_device      (GtkMenu             *menu,
                                            GdkDevice           *device,
-- 
2.4.0

